File: t/PACSFramework/CS/Leadtools.Medical.Winforms/UserAccess/Extensions.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;

#if (LEADTOOLS_V19_OR_LATER_MEDICAL_EXTERNAL_STORE) || (LEADTOOLS_V19_OR_LATER)
using System.ComponentModel;
#endif

namespace Leadtools.Medical.Winforms
{
   public static class Extensions
   {
      public static T Clone<T>(this T obj)
      {
         using (var ms = new MemoryStream())
         {
            var formatter = new BinaryFormatter();
            formatter.Serialize(ms, obj);
            ms.Position = 0;

            return (T)formatter.Deserialize(ms);
         }
      }

      public static void AddIfNotPresent<TKey, TValue>(this Dictionary<TKey, TValue> dict, TKey key, TValue value)
      {
         if (!dict.ContainsKey(key))
         {
            dict.Add(key, value);
         }
      }

#if (LEADTOOLS_V19_OR_LATER_MEDICAL_EXTERNAL_STORE) || (LEADTOOLS_V19_OR_LATER)
      public static string Description(this Enum enumValue)
      {
         var enumType = enumValue.GetType();
         var field = enumType.GetField(enumValue.ToString());
         var attributes = field.GetCustomAttributes(typeof(DescriptionAttribute), false);
         return attributes.Length == 0
             ? enumValue.ToString()
             : ((DescriptionAttribute)attributes[0]).Description;
      }
#endif

      public static string MyTrim(this string str)
      {
         if (string.IsNullOrEmpty(str))
         {
            return string.Empty;
         }

         return str.Trim();
      }
   }
}

File: t/PACSFramework/VB/StorageServerManager/DataTypes/Extensions.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary
Imports System.Security
Imports System.Reflection
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Runtime.InteropServices

Namespace Leadtools.Demos.StorageServer
   Public Class Extensions
      Private Sub New()
      End Sub

      Public Shared Function Clone(Of T)(ByVal obj As T) As T
         Using ms As MemoryStream = New MemoryStream()
            Dim formatter As BinaryFormatter = New BinaryFormatter()
            formatter.Serialize(ms, obj)
            ms.Position = 0

            Return CType(formatter.Deserialize(ms), T)
         End Using
      End Function

      Public Shared Function ToSecureString(ByVal password As String) As SecureString
         Dim passwordChars As Char()
         Dim secure As SecureString


         passwordChars = password.ToCharArray()
         secure = New SecureString()

         For Each c As Char In passwordChars
            secure.AppendChar(c)
         Next

         secure.MakeReadOnly()

         Return secure
      End Function

      Public Shared Function [Implements](Of T)(ByVal assembly As Assembly) As IEnumerable(Of Type)
         Dim itype As Type = GetType(T)

         Return assembly.GetTypes().Where(Function(type) itype.IsAssignableFrom(type))
      End Function

      Public Shared Sub UncheckAll(ByVal listview As ListView)
         For Each item As ListViewItem In listview.CheckedItems
            item.Checked = False
         Next item
      End Sub

      Public Shared Sub CheckAll(ByVal listview As ListView)
         For Each item As ListViewItem In listview.Items
            item.Checked = True
         Next item
      End Sub

      Public Shared Sub SetForeColor(ByVal listview As ListView, ByVal color As Color)
         For Each item As ListViewItem In listview.Items
            item.ForeColor = color
         Next item
      End Sub

      'Public Shared Sub Freeze(ByVal control As Control)
      '   Try
      '      SendMessage(control.Handle, WM_SETREDRAW, 0, IntPtr.Zero)
      '   Catch exception As Exception
      '      System.Diagnostics.Debug.Assert(False)

      '      Throw exception
      '   End Try
      'End Sub

      'Public Shared Sub UnFreeze(ByVal control As Control)
      '   Try
      '      SendMessage(control.Handle, WM_SETREDRAW, 1, IntPtr.Zero)
      '   Catch exception As Exception
      '      System.Diagnostics.Debug.Assert(False)

      '      Throw exception
      '   End Try
      'End Sub
   End Class
End Namespace
